File: src/fim_one/web/schemas/channel.py
"""Pydantic request/response schemas for the Channel API."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Create / Update
# ---------------------------------------------------------------------------


class ChannelCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    type: str = Field(pattern=r"^(feishu|wecom|slack)$")
    org_id: str = Field(min_length=1)
    config: dict[str, Any] = Field(default_factory=dict)
    is_active: bool = True


class ChannelUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    config: dict[str, Any] | None = None
    is_active: bool | None = None


# ---------------------------------------------------------------------------
# Responses
# ---------------------------------------------------------------------------


# Config fields considered sensitive — masked in API responses.
_SENSITIVE_KEYS = frozenset(
    {"app_secret", "verification_token", "encrypt_key", "api_secret"}
)


def _mask_config(config: dict[str, Any]) -> dict[str, Any]:
    """Return a shallow copy of ``config`` with sensitive values masked."""
    masked: dict[str, Any] = {}
    for key, value in config.items():
        if key in _SENSITIVE_KEYS and value:
            masked[key] = "***"
        else:
            masked[key] = value
    return masked


class ChannelResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    type: str
    org_id: str
    config: dict[str, Any]
    is_active: bool
    created_by: str
    created_at: str
    updated_at: str | None = None
    callback_url: str | None = None

    @classmethod
    def from_orm_masked(
        cls,
        channel: Any,
        *,
        callback_url: str | None = None,
    ) -> ChannelResponse:
        """Build a response with ``config`` sensitive fields masked."""
        return cls(
            id=channel.id,
            name=channel.name,
            type=channel.type,
            org_id=channel.org_id,
            config=_mask_config(dict(channel.config or {})),
            is_active=channel.is_active,
            created_by=channel.created_by,
            created_at=(
                channel.created_at.isoformat() if channel.created_at else ""
            ),
            updated_at=(
                channel.updated_at.isoformat() if channel.updated_at else None
            ),
            callback_url=callback_url,
        )


class ChannelListResponse(BaseModel):
    items: list[ChannelResponse]


class ChannelTestResponse(BaseModel):
    ok: bool
    error: str | None = None


# ---------------------------------------------------------------------------
# Hook Playground — test the real approval-gate flow end-to-end
# ---------------------------------------------------------------------------


class TestApprovalRequest(BaseModel):
    """Body for ``POST /api/channels/{id}/test-approval``.

    All fields are optional — the endpoint fills sensible defaults modeled
    on a realistic "sensitive tool call" so demos look natural out of the
    box.
    """

    tool_name: str | None = Field(
        default=None,
        max_length=200,
        description="Simulated tool name shown on the approval card.",
    )
    tool_args: dict[str, Any] | None = Field(
        default=None,
        description="Simulated tool arguments (rendered as pretty JSON).",
    )
    title: str | None = Field(default=None, max_length=200)
    summary: str | None = Field(default=None, max_length=2000)


class TestApprovalResponse(BaseModel):
    ok: bool
    confirmation_id: str | None = None
    error: str | None = None


class ConfirmationStatusResponse(BaseModel):
    """Current state of a ``ConfirmationRequest`` row — polled by the UI."""

    id: str
    status: str  # pending | approved | rejected | expired
    tool_name: str | None = None
    tool_args: dict[str, Any] | None = None
    test_mode: bool = False
    created_at: str
    responded_at: str | None = None
    responded_by_open_id: str | None = None


# ---------------------------------------------------------------------------
# Chat discovery (Feishu group picker)
# ---------------------------------------------------------------------------


class ChatDiscoveryRequest(BaseModel):
    """Request to list Feishu groups that the app/bot is a member of.

    The caller passes ``app_id`` + ``app_secret`` when creating a channel
    (no DB row exists yet).  When editing an existing channel, the caller
    may pass ``channel_id`` alone — the server will decrypt the stored
    ``app_secret`` to query Feishu.  If both ``app_secret`` and
    ``channel_id`` are present, ``app_secret`` wins (user explicitly
    re-entered it).
    """

    app_id: str = Field(min_length=1)
    app_secret: str | None = None
    channel_id: str | None = None
    org_id: str | None = None


class ChatInfo(BaseModel):
    chat_id: str
    name: str
    avatar: str | None = None
    description: str | None = None
    member_count: int | None = None
    external: bool = False


class ChatDiscoveryResponse(BaseModel):
    items: list[ChatInfo]


__all__ = [
    "ChannelCreate",
    "ChannelUpdate",
    "ChannelResponse",
    "ChannelListResponse",
    "ChannelTestResponse",
    "TestApprovalRequest",
    "TestApprovalResponse",
    "ConfirmationStatusResponse",
    "ChatDiscoveryRequest",
    "ChatInfo",
    "ChatDiscoveryResponse",
]

File: frontend/src/types/channel.ts
/**
 * Messaging channel types.
 *
 * A channel connects an external messaging platform (Feishu, WeCom, Slack, …)
 * to the FIM One agent runtime. Agents can push notifications and receive
 * interactive actions (e.g. card button clicks) through the configured channel.
 *
 * Currently only `feishu` is wired up; `wecom`/`slack`/`teams` are reserved
 * for future releases.
 */

export type ChannelType = "feishu" | "wecom" | "slack" | "teams"

/**
 * Sanitised configuration returned by the backend.
 *
 * Secret fields (`app_secret`, `verification_token`, `encrypt_key`) are never
 * echoed back — instead the API returns `*_configured` booleans so the UI can
 * show a "configured" badge or a "leave blank to keep" hint.
 */
export interface ChannelConfig {
  // Feishu fields
  app_id?: string
  chat_id?: string
  chat_name?: string
  app_secret_configured?: boolean
  verification_token_configured?: boolean
  encrypt_key_configured?: boolean

  // Reserved for future channel types
  [key: string]: unknown
}

export interface Channel {
  id: string
  name: string
  type: ChannelType
  org_id: string
  config: ChannelConfig
  /** Canonical callback URL that the third-party platform should POST to. */
  callback_url: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ChannelListResponse {
  items: Channel[]
}

/**
 * Request bodies.
 *
 * For create/update, clients send the plain-text secrets. On update, an empty
 * string means "do not touch the stored value".
 */
export interface FeishuChannelConfigInput {
  app_id: string
  app_secret?: string
  chat_id: string
  verification_token?: string
  encrypt_key?: string
}

export interface ChannelCreateRequest {
  name: string
  type: ChannelType
  org_id: string
  config: FeishuChannelConfigInput | Record<string, unknown>
}

export interface ChannelUpdateRequest {
  name?: string
  config?: FeishuChannelConfigInput | Record<string, unknown>
  is_active?: boolean
}

export interface ChannelTestResponse {
  ok: boolean
  error?: string
  chat_name?: string
}

/**
 * Hook Playground — exercises the real FeishuGateHook round-trip from the UI.
 *
 * Unlike `/test` (which sends a preview card with a sentinel id whose button
 * clicks are no-ops), `test-approval` creates a genuine `ConfirmationRequest`
 * row.  The UI then polls `getConfirmation()` until the status flips.
 */
export interface TestApprovalRequest {
  tool_name?: string
  tool_args?: Record<string, unknown>
  title?: string
  summary?: string
}

export interface TestApprovalResponse {
  ok: boolean
  confirmation_id?: string
  error?: string
}

export type ConfirmationStatusValue =
  | "pending"
  | "approved"
  | "rejected"
  | "expired"

export interface ConfirmationStatus {
  id: string
  status: ConfirmationStatusValue
  tool_name?: string | null
  tool_args?: Record<string, unknown> | null
  test_mode: boolean
  created_at: string
  responded_at?: string | null
  responded_by_open_id?: string | null
}

/**
 * Payload for `POST /api/channels/discover-chats`.
 *
 * - **Create mode**: provide `app_id` + `app_secret` + `org_id`.
 * - **Edit mode**: provide `channel_id`; the server reuses the stored
 *   (encrypted) secret. Pass `app_secret` only if the user re-typed it.
 */
export interface ChatDiscoveryRequest {
  app_id: string
  app_secret?: string
  channel_id?: string
  org_id?: string
}

export interface ChatInfo {
  chat_id: string
  name: string
  avatar?: string | null
  description?: string | null
  member_count?: number | null
  external: boolean
}

export interface ChatDiscoveryResponse {
  items: ChatInfo[]
}
